builtin/logical/aws/rotation_test.go

```diff
@@ -0,0 +1,196 @@
+						Username:       "jane-doe",
+						RotationPeriod: 1 * time.Minute,
+					},
+					changed: false,
+				},
+			},
+		},
+		{
+			name: "refresh one of two",
+			creds: []credToInsert{
+				{
+					config: staticRoleConfig{
+						Name:           "test",
+						Username:       "jane-doe",
+						RotationPeriod: 1 * time.Minute,
+					},
+					changed: false,
+				},
+				{
+					config: staticRoleConfig{
+						Name:           "toast",
+						Username:       "john-doe",
+						RotationPeriod: 1 * time.Second,
+					},
+					changed: true,
+				},
+			},
+		},
+	}
+
+	// wrap our cred generating function so we don't have to handle the error _every time_.
+	// we pass in the T so that we have the correct subtest T for each call.
+	randStr := func(len int, t *testing.T) *string {
+		str, err := credsutil.RandomAlphaNumeric(len, false)
+		if err != nil {
+			t.Fatalf("couldn't generate a random string for a key: %s", err)
+		}
+		return aws.String(str)
+	}
+
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			config := logical.TestBackendConfig()
+			config.StorageView = &logical.InmemStorage{}
+
+			b := Backend()
+			err := b.initQueue(context.Background(), nil)
+			if err != nil {
+				t.Fatalf("couldn't initialize queue: %s", err)
+			}
+
+			// this means the creds will be the same for every user, but that's okay
+			// since what we care about is whether they changed on a single-user basis.
+			miam, err := awsutil.NewMockIAM(
+				// blank list for existing user
+				awsutil.WithListAccessKeysOutput(&iam.ListAccessKeysOutput{
+					AccessKeyMetadata: []*iam.AccessKeyMetadata{
+						{},
+					},
+				}),
+				// initial key to store
+				awsutil.WithCreateAccessKeyOutput(&iam.CreateAccessKeyOutput{
+					AccessKey: &iam.AccessKey{
+						AccessKeyId:     randStr(16, t),
+						SecretAccessKey: randStr(16, t),
+					},
+				}),
+			)(nil)
+
+			// insert all our creds
+			for i, cred := range c.creds {
+
+				if err != nil {
+					t.Fatalf("couldn't initialze mock IAM handler: %s", err)
+				}
+				b.iamClient = miam
+
+				err = b.createCredential(context.Background(), config.StorageView, cred.config)
+				if err != nil {
+					t.Fatalf("couldn't insert credential %d: %s", i, err)
+				}
+
+				item := &queue.Item{
+					Key:      cred.config.Name,
+					Value:    cred.config,
+					Priority: time.Now().Add(cred.config.RotationPeriod).Unix(),
+				}
+				err = b.credRotationQueue.Push(item)
+				if err != nil {
+					t.Fatalf("couldn't push item onto queue: %s", err)
+				}
+			}
+
+			time.Sleep(5 * time.Second)
+
+			// update aws responses, same argument for why it's okay every cred will be the same
+			miam, err = awsutil.NewMockIAM(
+				// old key
+				awsutil.WithListAccessKeysOutput(&iam.ListAccessKeysOutput{
+					AccessKeyMetadata: []*iam.AccessKeyMetadata{
+						{
+							AccessKeyId: randStr(16, t),
+						},
+					},
+				}),
+				// new key - one char longer, so we guarantee it _changes_
```

Comment:
yeah, i think this case might need some rethought.

One thing we could do is have each case declare its old/new creds, but i was trying to avoid that because theoretically the inputter doesn't have control over that information.

The other thing is that we could use the same old/new secret (hardcoded) for every cred, since they theoretically could be the same. This removes the randomness, and the vanishingly small chance that the old/new secret keys could collide.

Reply:
I rewrote the test eliminating the randomness. Every cred gets the same secret (but different, deterministic, AccessKeyID), and rotating keys get the same "new" secret, and we check that.